From 164d4de5b26f96d0423f7172a8560d1d70ece235 Mon Sep 17 00:00:00 2001
From: Robby Cai <robby.cai@nxp.com>
Date: Wed, 23 May 2018 21:03:27 +0800
Subject: [PATCH 3878/5242] MLK-18362-2 midea: mipi_csi: add mipi phy reset on
 imx8mm

commit  0e9049979e58304d4979697278dbe2e6ad533c1d from
https://source.codeaurora.org/external/imx/linux-imx.git

on imx8mm, use different way rather than via SRC. so use different
device id to distinguish different MIPI-PHY reset method.

Signed-off-by: Robby Cai <robby.cai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/mxc/capture/mxc_mipi_csi.c |   40 +++++++++++++++++++--
 1 file changed, 38 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/mxc/capture/mxc_mipi_csi.c b/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
index 483519b..d174147 100644
--- a/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
+++ b/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
@@ -33,6 +33,7 @@
 #include <linux/mfd/syscon.h>
 #include <linux/module.h>
 #include <linux/of.h>
+#include <linux/of_address.h>
 #include <linux/of_graph.h>
 #include <linux/platform_device.h>
 #include <linux/pm_runtime.h>
@@ -189,6 +190,8 @@
 #define MIPI_CSIS_PKTDATA_EVEN		0x3000
 #define MIPI_CSIS_PKTDATA_SIZE		SZ_4K
 
+#define MIPI_CSIS_MISC			0x8008
+
 #define DEFAULT_SCLK_CSIS_FREQ	166000000UL
 
 enum {
@@ -328,6 +331,8 @@ struct csis_pix_format {
 	}
 };
 
+typedef int (*mipi_csis_phy_reset_t)(struct csi_state *state);
+
 #define mipi_csis_write(__csis, __r, __v) writel(__v, __csis->regs + __r)
 #define mipi_csis_read(__csis, __r) readl(__csis->regs + __r)
 
@@ -383,6 +388,25 @@ static int mipi_csis_phy_init(struct csi_state *state)
 	return ret;
 }
 
+static int mipi_csis_phy_reset_mx8mm(struct csi_state *state)
+{
+	struct device_node *np;
+	void __iomem *reg;
+
+	np = of_find_compatible_node(NULL, NULL, "fsl,imx8mm-csi");
+	if (WARN_ON(!np))
+		return -ENXIO;
+
+	reg  = of_iomap(np, 0);
+
+	writel(0, reg + MIPI_CSIS_MISC);
+	usleep_range(10, 20);
+	writel(0x30000, reg + MIPI_CSIS_MISC);
+	usleep_range(10, 20);
+
+	return 0;
+}
+
 static int mipi_csis_phy_reset(struct csi_state *state)
 {
 	struct device_node *np = state->dev->of_node;
@@ -1026,6 +1050,8 @@ static int mipi_csis_probe(struct platform_device *pdev)
 	struct v4l2_subdev *mipi_sd;
 	struct resource *mem_res;
 	struct csi_state *state;
+	const struct of_device_id *of_id;
+	mipi_csis_phy_reset_t phy_reset_fn;
 	int ret = -ENOMEM;
 
 	state = devm_kzalloc(dev, sizeof(*state), GFP_KERNEL);
@@ -1050,7 +1076,12 @@ static int mipi_csis_probe(struct platform_device *pdev)
 	}
 
 	mipi_csis_phy_init(state);
-	mipi_csis_phy_reset(state);
+	of_id = of_match_node(mipi_csis_of_match, dev->of_node);
+	if (!of_id || !of_id->data)
+		return -EINVAL;
+
+	phy_reset_fn = of_id->data;
+	phy_reset_fn(state);
 
 	mem_res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	state->regs = devm_ioremap_resource(dev, mem_res);
@@ -1225,7 +1256,12 @@ static int mipi_csis_remove(struct platform_device *pdev)
 };
 
 static const struct of_device_id mipi_csis_of_match[] = {
-	{	.compatible = "fsl,imx7d-mipi-csi",},
+	{	.compatible = "fsl,imx7d-mipi-csi",
+		.data = (void *)&mipi_csis_phy_reset,
+	},
+	{	.compatible = "fsl,imx8mm-mipi-csi",
+		.data = (void *)&mipi_csis_phy_reset_mx8mm,
+	},
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, mipi_csis_of_match);
-- 
1.7.9.5

